Allow parameters in CREATE/ALTER DATABASE TOPOLOGY (#1911)
neo4j/docs-operations: modules/ROOT/pages/clustering/databases.adoc

----
CREATE DATABASE foo TOPOLOGY 3 PRIMARIES 2 SECONDARIES
----

> Alternatively, using link:{neo4j-docs-base-uri}/cypher-manual/{page-version}/syntax/parameters[parameters]:

Reviewer: ```suggestion
Alternatively, starting from 5.26, you can use link:{neo4j-docs-base-uri}/cypher-manual/{page-version}/syntax/parameters[parameters] to provide the number of primaries and secondaries:
```

At this point you can run `REALLOCATE DATABASES` to make the cluster re-balance databases across all servers that are part of the cluster.
====

> The `CREATE DATABASE` and `ALTER DATABASE` commands support the usage of link:{neo4j-docs-base-uri}/cypher-manual/{page-version}/syntax/parameters[parameters] when providing the number of primaries and secondaries.

Reviewer 2: I wonder if this is overkill. We don't mention that the name could be a parameter. Maybe enough to add a second example to the create example. Something like:

```cypher
CREATE DATABASE foo TOPOLOGY 3 PRIMARIES 2 SECONDARIES
```

or with parameters

```cypher
{
  "dbname": "foo",
  "primary": 1,
  "secondary": 2
}

CREATE DATABASE $dbName TOPOLOGY $primary PRIMARIES $secondary SECONDARIES

Reviewer: I think it's useful to have both.

Author: Added the dbname 👍 

I feel like it would be overkill to have one parameter example for create and another for alter alter, by mentioning ALTER in the text we get away with only the one. Let me know if you'd prefer the `or with parameters` approach with one for create and one for alter 🤷

Reviewer: I think a better practice is to keep them separate because whoever is looking for how to alter a database won't be looking at the create a database section. Of course, they might find a way to the example, but nobody reads the docs from top to bottom. A workaround could be to mention the feature, from when it is supported, etc., in the alter section and then refer to the create section, but it might be even more complicated.

At this point you can run `REALLOCATE DATABASES` to make the cluster re-balance databases across all servers that are part of the cluster.
====

> The `CREATE DATABASE` and `ALTER DATABASE` commands support the usage of link:{neo4j-docs-base-uri}/cypher-manual/{page-version}/syntax/parameters[parameters] when providing the database, the number of primaries and/or the number of secondaries.

Reviewer: ```suggestion
The `CREATE DATABASE` and `ALTER DATABASE` commands support the usage of link:{neo4j-docs-base-uri}/cypher-manual/{page-version}/syntax/parameters[parameters] when providing the database, the number of primaries, and/or the number of secondaries.
```

Reviewer: Missing Oxford comma.

Author: No longer applicable, I think

----
ALTER DATABASE foo SET TOPOLOGY 2 PRIMARIES 1 SECONDARY
----

> Alternatively, using link:{neo4j-docs-base-uri}/cypher-manual/{page-version}/syntax/parameters[parameters]:

Reviewer: ```suggestion
Alternatively, starting from 5.26, you can use link:{neo4j-docs-base-uri}/cypher-manual/{page-version}/syntax/parameters[parameters] to provide the number of primaries and secondaries:
```

`ALTER DATABASE` commands are optionally idempotent, with the default behavior to fail with an error if the database does not exist.
Appending `IF EXISTS` to the command ensures that no error is returned and nothing happens should the database not exist.

> If the `ALTER DATABASE` command decreases the number of allocations of a database, allocations on xref:clustering/servers.adoc#cordoned_state[cordoned servers] are removed first.

Reviewer: ```suggestion
If the `ALTER DATABASE` command decreases the number of allocations of a database, allocations on xref:clustering/servers.adoc#cordoned-state[cordoned servers] are removed first.
```

The topology of `foo` is stored in the `system` database and when you create it, it is allocated according to the default topology (which can be shown with `CALL dbms.showTopologyGraphConfig`).
This may be different from the topology of `foo` when it was backed up.
If you want to ensure a certain allocation across the cluster, you can specify the desired topology with the `TOPOLOGY` clause in the `CREATE DATABASE` command.

> See <<#create-database, `CREATE DATABASE`>> for more information.

Reviewer: This should be working without the # tag. Good catch, btw.
```suggestion
See <<create-database, `CREATE DATABASE`>> for more information.
```
